freeipa/src/main/java/com/sequenceiq/freeipa/converter/stack/StackToCreateFreeIpaRequestConverter.java

```diff
@@ -0,0 +1,393 @@
+import com.sequenceiq.freeipa.entity.SecurityRule;
+import com.sequenceiq.freeipa.entity.Stack;
+import com.sequenceiq.freeipa.entity.StackAuthentication;
+import com.sequenceiq.freeipa.entity.Template;
+import com.sequenceiq.freeipa.service.freeipa.FreeIpaService;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.core.convert.converter.Converter;
+import org.springframework.stereotype.Component;
+
+import javax.inject.Inject;
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.nio.file.Paths;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+@Component
+public class StackToCreateFreeIpaRequestConverter implements Converter<Stack, CreateFreeIpaRequest> {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(StackToCreateFreeIpaRequestConverter.class);
+
+    private static final String DELETED_NAME_DELIMETER = "_";
+
+    private static final String PATH_DELIMETER = "/";
+
+    private static final String CLUSTER_BACKUP_PREFIX = "cluster-backups";
+
+    private static final String CLUSTER_LOG_PREFIX = "cluster-logs";
+
+    private static final String CLUSTER_TYPE = "freeipa";
+
+    private static final String AZURE_BLOB_STORAGE_SCHEMA = "https";
+
+    private static final String ORIGINAL_AZURE_BLOB_STORAGE_SCHEMA = "abfs://";
+
+    @Inject
+    private FreeIpaService freeIpaService;
+
+    @Override
+    public CreateFreeIpaRequest convert(Stack source) {
+        CreateFreeIpaRequest request = new CreateFreeIpaRequest();
+
+        request.setEnvironmentCrn(source.getEnvironmentCrn());
+        request.setName(getName(source));
+        request.setPlacement(getPlacementRequest(source));
+        request.setInstanceGroups(source.getInstanceGroups().stream().map(this::getInstanceGroupRequst).collect(Collectors.toList()));
+        request.setAuthentication(getStackAuthenticationRequest(source.getStackAuthentication()));
+        request.setNetwork(getNetworkRequest(source.getNetwork()));
+        request.setImage(getImageSettingsRequest(source.getImage()));
+        request.setFreeIpa(getFreeIpaServerRequest(source));
+        request.setGatewayPort(source.getGatewayport());
+        request.setTelemetry(getTelemetry(source));
+        request.setBackup(getBackup(source));
+        request.setTags(getTags(source.getTags()));
+        request.setUseCcm(source.getUseCcm());
+        request.setTunnel(source.getTunnel());
+        request.setVariant(source.getPlatformvariant());
+
+        return request;
+    }
+
+    private String getName(Stack stack) {
+        String name = stack.getName();
+        if (stack.isDeleteCompleted()) {
+            int lastDelimeterIndex = name.lastIndexOf(DELETED_NAME_DELIMETER);
+            if (lastDelimeterIndex != -1) {
+                name = name.substring(0, lastDelimeterIndex);
+            }
+        }
```

Comment:
if the stack is not deleted properly, then it's not renamed. In that case a new stack would be created with the same name which would fail on the database side as there are unique constraints for name+envcrn+accountid